### ISSUE TITLE
Partitions gets unhealthy if many timers are scheduled

### ISSUE THREAD

**Issue by ChrisKujawa** [camunda/camunda#8991]
**Describe the bug**

We deployed the following process

![gamedayprocess](https://user-images.githubusercontent.com/2758593/160121847-f36619db-f568-46b0-a6f9-8caa80f68f58.png)

There were two bugs in the process:

 *  an parallel sequence flow to the call activity, causing endless recursion
 * the timer expression was not optimal `=date and time(today(), time("16:00:00@Europe/Berlin")) - now()` it caused an negative duration after reaching the time.

When we deploy the model, we create endless new process instances and fill the timer column. If the specific time is reached we have to trigger a lot of timers, might be millions depending how long the recursion was running and how many instances have initially be started. The recursion continues and adds more and more due timers to the timer column.

[The problem here is that the processing is completely blocked, since the timer checker is running in the same actor.](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/timer/DueDateTimerChecker.java). Furthermore, it causes to produces an endless backlog, which the processor is not able to catch up. The processor will become unhealthy and the backpressure wents up to 100% since the requests are not answered. 

![general](https://user-images.githubusercontent.com/2758593/160123440-bf01adff-6b71-4578-bdef-e162f059373a.png)
![processing](https://user-images.githubusercontent.com/2758593/160123637-706cba79-53c0-437b-a8f6-e08d0a83b53f.png)

**To Reproduce**

Deploy this model 
[gamedayprocess.bpmn.txt](https://github.com/camunda/zeebe/files/8350535/gamedayprocess.bpmn.txt) and start new instances watch out, you can throw your cluster away afterwards. 

**Expected behavior**
As a user I'm not able to completely f** up my cluster, or at least there is some way to recover from this.

**Comment by npepinpe** [comment]
Seems like we did not triage this, and this was marked as critical. Let's talk about it tomorrow - I guess the board switch caused me to miss this one.

**Comment by pihme** [comment]
This also occurred in a simpler process. E.g. in Roman's benchmark he wants to have a certain load on the system. He starts a number of process instance e.g. 10/s. Each process waits at two points for 20 minutes. Because of the continuous creation of process instances, at any point in time, there are a couple of timers ready to be triggered. 

Depending on the load the performance is heavily degraded.

**Comment by pihme** [comment]
Roman's workflow looked like this:
![image](https://user-images.githubusercontent.com/14032870/165306953-8b862363-59ee-4dfd-9533-63f2fba6a88c.png)

I guess it can be simplified to:
![image](https://user-images.githubusercontent.com/14032870/165307116-4942fb88-823d-48ac-b952-a72859d6bdca.png)

But maybe the other elements are necessary to add latency between the time a command is written to the log and the time a record is processed.

This is just to say it can also happen without specially crafted process from Game days.

Looking at the code I can see the following:
* Timer Records are added to the database when the timer element is activated
* Timer records are removed after the `TimerIntent.TRIGGER` has been processed.

This also means that the `DueDateTimeChecker` might visit the same timer multiple times, and create multiple trigger commands for it. This is because the `DueDateTimeChecker` does not alter the database, and only writes to the log.

So my working hypothesis is that the there is an escalating feedback loop.
* `DueDateTimeChecker` visits a couple of timers once, writes commands for these timers
* `DueDateTimeChecker` is triggered again before these commands have been processed, thus revisit the same events and also new events which have come due since the last run. Then it writes duplicate commands for the old timers and new commands for new timers it found
* Thus, if we cannot process all commands than the `DueDateTimeChecker` wrote in the last run, before `DueDateTimeChecker` is running again, we will write more and more duplicate records to the commit log and revisit certain timers again and again.
* Sine the db table is chronological, the commands are also chronological. So in the second round we encounter the duplicate commands first and waste time on those, which means we make even less progress in terms of removing the new events from the database
* This can then escalate until we are totally blocked, because thousands of timers have not been removed yet, and thousands of timers lead to thousands of follow up records

Could also be that this happens more prominently after a broker resumes, because much time has passed since the last run. This could also explain Romans observation that one some days he can get stable performance and on others not.

/cc @Zelldon @saig0  wdyt?

**Comment by saig0** [comment]
@pihme yes. That sounds correct. 

But there is one more aspect. If the `DueDateTimeChecker` visits many timers on a check then it blocks the actor thread and other actor jobs can't be executed (e.g. sending heartbeats). 

A solution could be to limit the number of timers for one check (e.g. only 100 timers at once).

**Comment by pihme** [comment]
Not sure this would be a valid solution. My fear is that it will just always be falling further and further behind. 

Right now the highest possible frequency of `DueDateTimeChecker` runs seems to be 10 Hertz. If we only allow to trigger 100 timers per invocation, but we create 200 timers each 100ms, then we will just fall further and further behind. And this build up would not be reflected in the back pressure mechanism.

My hope is that if we always trigger all due timers, this will cause delay, which will influence back pressure. The way the system is set up now, I see no  other way than to let back pressure dial in to sustained throughput to the speed of the slowest link in the processing chain.

**Comment by Zelldon** [comment]
Sorry for the late response. Interesting observation you made @pihme and sounds right. In my case it was more like @saig0 described that the actor didn't stop to write due timers, I had many many write records per second but no processing at all. 

I agree with the argument from @pihme. Not sure whether it would really help to just limit it, sounds like a band-aid :adhesive_bandage: I guess optimal would be to have a separate actor, which writes to the log. This would at least help here with not blocking processing. Still it can happen that we write many many timers at once. We could solve that then maybe via writing over the command API or something such that back pressure is taken into account like @pihme said. :shrug:

**Comment by pihme** [comment]
@Zelldon Could you comment on the prioritization of actor tasks with respect to time based tasks vs. ordinary tasks.

More precisely: Supposing the `DueDateTimeChecker` is scheduled every 100ms, but runs more than 100ms each time. What will happen? Will the actor schedule always and only schedule the `DueDateTimeChecker` or will some progress be made by other tasks before `DueDateTimeChecker` is scheduled again?

**Comment by pihme** [comment]
And the second question: What happens when you schedule a task with a negative delay? Because I think this can happen

**Comment by pihme** [comment]
- [ ] asad

**Comment by Zelldon** [comment]
Hey @pihme 

as always it depends. In general, the timers and "ordinary" tasks are at the end scheduled the same. We have the fastlane and the queue, for "less" priority. So we have kind of an implicit priority. What I understand from the code is that if there are jobs in the fastlane already they are executed first and due timers are [ added to the fastlane as well](https://github.com/camunda/zeebe/blob/dd3e9fb653e04223e4fe64b88b3273231efa6c6e/util/src/main/java/io/camunda/zeebe/util/sched/ActorTask.java#L393). I guess the issue is even more problematic if we use `runAtFixedRate` and the execution takes longer than the rate. 

To get to your example: If the `DueDateTimeChecker` is scheduled via `runAtFixedRate` with 100ms as the rate and takes longer than `100ms` then a new job will already be added. Normally we use `runDelayed`, which means we schedule the next job only after we are done.[ AFAIK this is how it's done in the `DueDateTimeChecker`.](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/scheduled/DueDateChecker.java#L47-L56)

So in the above-described example and scenario, which I used in the gameday was that the DueTimerchecker had thousand or millions of Timers, which are due at the same time. It blocked the actor thread completely because it iterated over the whole column family to write a timer trigger command. I guess due to the subscription logic and "priority" that timers are scheduled in fast-lane the processing was not happening for a long time. This is my current assumption.

Regarding the negative duration, this happened to me accidentally in the gameday. This means the timer is immediately due.

**Comment by pihme** [comment]
## Summary of Preliminary Findings

There are two problems we need to solve:
* Huge backlog of accumulated timers
* Steady state with high throughput of process instances

The first problem was the focus of the original description of the issue. This could for example occur when a company wants to send their customers a thank you email on Christmas. Throughout the year timers are collected, which all fire on Christmas day simultaneously. 

Here we see the following issues:
* [ ] Timers are iterated over in full without ever giving yield. During this time, no other processing can take place

The second problem was observed by Roman in benchmarks. Here it is just a steady load which leads to indeterministic blocking of the processing thread.

Here we see the following issues:
* [ ] Due timers may be visited several times by the `DueDateTimeChecker`. Each time it will write a new `TriggerTimer` command. Only when the command is processed will the timer be removed. So then we run into a difference in the speed in which timers are created, and the speed in which timers are removed. And writing duplicate records makes this worse for reasons outline above.
* [ ] The `DueDateTimeChecker` may be scheduled with a negative delay to run next, which means it runs pretty much immediately.

**Comment by romansmirnov** [comment]
Just want to share my thoughts 😄 

@Zelldon,

Maybe a comment on your https://github.com/camunda/zeebe/issues/8991#issuecomment-1110603307 😄 

> I guess due to the subscription logic and "priority" that timers are scheduled in fast-lane the processing was not happening for a long time. This is my current assumption.

In the described scenario, with millions of timers being due simultaneously, I think the problem is not with how the actor schedules the jobs. In my opinion, this `whileTrue`-loop is the problem:

https://github.com/camunda/zeebe/blob/dd3e9fb653e04223e4fe64b88b3273231efa6c6e/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbTimerInstanceState.java#L90-L105

This means it will remain in this loop as long as there is a next timer event being due. With millions of timers being due simultaneously, it will execute that loop until it visits all the millions of timer events...and this may take some time and block processing on that partition in the meantime.

@pihme,

> Here it is just a steady load which leads to indeterministic blocking of the processing thread.

I don't think my scenario is much different from Chris's scenario in the gameday. The only difference is there are more than 20k timer events on one partition, whereby the timer's due date is distributed over the next 20 minutes. So, not all of them are due at the same time.

What can happen (and happens) is that it will remain in the `whileTrue`-loop for the next 20 minutes because there is always a next timer event being due. Also, it may happen that in between, it reaches a moment when the next timer event is not due at that moment (but the timer event would get due at the very next moment). Then it will schedule the next run accordingly, but because of how the actor schedules the jobs (as described by @Zelldon), it will immediately go into the `whileTrue`-loop and continue visiting the timers.

**Comment by romansmirnov** [comment]
> I guess optimal would be to have a separate actor, which writes to the log.

I agree with that point. I even tried that but it failed with the exception in #9187 - but I haven't investigated it further.

Besides that, the only concern I may have, if there is only one CPU thread configured 😉

**Comment by Zelldon** [comment]
> Besides that, the only concern I may have, if there is only one CPU thread configured wink

@romansmirnov you mean because it would block the thread again and the other actor can't run? Might be then the yield or backpressure would help here? :thinking:

**Comment by pihme** [comment]
> In the described scenario, with millions of timers being due simultaneously,

In your case @romansmirnov the question in my mind is: How did you get to the millions of timers being due? 

If all runs smoothly, you create 10 PI/s per second. Even if you have just one partition this makes 10 new timers due per second in one partition. The checker runs also 10 times a second. So each time it runs, 
* it takes the threshold due time before entering the loop, 
* it would find 1 due timer before the threshold due time, 
* then in the next iteration of the loop it would find a timer that is after the threshold time
* It takes this timers' due time to reschedule itself then exists the loop

To get to millions of due timers, it means that it must have either missed 1 mio runs, or made no progress in removing timers for a long time.

I agree that once you get to this point, you are in the same boat as Chris (who had a bad timer expression and an endless recursion to build up that backlog). 

But how did you get there in the first place? That's what I am wondering.
Could be #9236
Could be https://github.com/camunda/zeebe/issues/8991#issuecomment-1109828994
Could be something else

**Comment by romansmirnov** [comment]
@pihme,

> In your case @romansmirnov the question in my mind is: How did you get to the millions of timers being due?

I didn't end up having millions of timers. With millions of timers, I was referring to Chris's https://github.com/camunda/zeebe/issues/8991#issuecomment-1110603307:

> ... had thousand or millions of Timers, which are due at the same time

In my scenario, I have > 65k timers in total (and roughly >20k timers per partition) whereby the timer's due dates are distributed over the next 20 minutes.

**Comment by pihme** [comment]
Update:

The following changes were implemented:
* timer will give yield after some time (currently 50ms)
* Scheduling of due date checker will leave some time (currently 100ms) between invocations

Expectation was that this should alleviate the problem reported by Chris (but not the one reported by Roman).

However, that was not the case.

Test setup: 
* Start Chris's process around 2:30 PM
* Wait till 4 PM
* Try to cancel the process, or deploy a new version which does not include recursion

Observation
* All requests to the Broker time out
* Even when just starting the Broker on its own, it will be unhealthy most of the time
* Sometimes it shortly switches to healthy, just to go back to unhealthy a couple of seconds later

Further research is needed.

I uploaded my data folder and config here: https://drive.google.com/file/d/165ysQ_Wv6iR2pIdsEXkA_AZchPdpvhVu/view?usp=sharing

**Comment by pihme** [comment]
I was able to deploy a process after doing a spike in `DueDateChecker` to prevent writing duplicate trigger timer commands.

Even with that spike, it took ~ 1h for the system to process the back log of records.

**Comment by pihme** [comment]
Update before handover:

There were three problems found in `DueDateChecker`
* Tasks to check for due timers was sometimes rescheduled immediately - fixed #9236
* DueDateTimeChecker did not yield - fixed #9238 
* DueDateTimeChecker will visit the same timer more than once and write duplicate trigger event records - open (no issue created yet)

With the two fixes, I was unable to repeatedly recover a cluster which was unresponsive. I was able to recover it once, by also implementing a quick fix for the third issue, but I was unable to reproduce the success.

There were a couple of problems found unrelated to `DueDateChecker`
* #9276 In discussion. Behavior seems to be expected, but is not really useful
* https://github.com/camunda/zeebe/issues/9283 open. Could help in this case and would be an easy fix

Another issue not needed for the fix, but to make it better to diagnose issue:
https://github.com/camunda/zeebe/issues/9282

From what I could see from the logs of the server it seems to be stuck at some other place as well. It makes good progress, intermitted by long breaks. I don't know what is causing the breaks. The data set is linked in one of the comments above.
I did add timer watch guards around the call to process a command:

```
        final long now = System.currentTimeMillis();
        processCommand(currentRecord);
        final long elapsedTime = System.currentTimeMillis() - now;
        if (elapsedTime > 1000) {
          System.out.println("Elapsed Time:" + Duration.ofMillis(elapsedTime).toString());
        }
```

However, I have not  observed any commands that exceeded 1 s.

Another thing to consider is whether the same issues of DueDateTimerChecker would also need to be implemented for job timeouts. (in the logs I saw a lot of jobs timing out)

**Comment by saig0** [comment]
## Summary

A partition gets unhealthy if many timers are scheduled for the same time (or in a short time frame).
The number of timers depends on the available resources. 

@romansmirnov ran into this issue with a process that has two timers after each other (including some service tasks in between). The timers are scheduled for 20 minutes.    

With cluster plan `S`: if more than 17 instances are created per second
With cluster plan `M`: if more than 27 instances are created per second

The partition got unhealthy but it recovered after a while (20 minutes - 1 hour).

-----

We will not work on this issue immediately because 
* it is not likely to have so many timers that are scheduled for the same time
* it is a general issue about processing large data from the database (also for job time outs and message TTL) that requires a bigger effort to fix it

**Comment by korthout** [comment]
Updated severity to reflect current team opinion described [here](https://github.com/camunda/zeebe/issues/8991#issuecomment-1127237068). Issue is moved into backlog.

**Comment by npepinpe** [comment]
@eppdot - any reason you put it back on our board?

**Comment by marcosgvieira** [comment]
### 🧭 Triage Assessment

| Field | Value |
|------|-------|
| **Work type** | BUG |
| **Severity** | 🔥 HIGH |
| **Likelihood** | ⚠️ Mid |
| **Confidence** | 0.95 |

---

### 📝 Rationale

The issue describes incorrect behavior in a released product: a partition becomes unhealthy and processing is blocked when many timers are scheduled simultaneously, with no simple workaround and significant impact on production usage. The problem is reproducible and has been observed in both synthetic and real-world scenarios. However, it is not considered critical because it does not cause data loss or security issues, and recovery is possible after some time. The likelihood is MID, as it requires specific (but not rare) process models or load patterns to trigger, and is not expected to affect all users frequently, but can occur in realistic high-load situations. The issue is clearly within Core Features team boundaries, and the report contains sufficient detail for triage.

---

🤖 _This issue was automatically triaged by an AI-based system as part of the Core Features triage workflow._

✋ **Want a human review?**
If you believe this assessment is incorrect or needs adjustment, feel free to update the details or request a manual triage by adding the **`triage:manual`** label.

**Comment by mustafadagher** [comment]
@ChrisKujawa It looks like you added this issue to Reliability testing board today, can we remove this from core-features board for now?

**Comment by ChrisKujawa** [comment]
@mustafadagher no this is just for tracking purposes it doesnt mean we are working on it

**Comment by marcosgvieira** [comment]
🔄 **Status updated automatically (Core Features)**

**From:** Backlog  
**To:** Ready  

**Reason:** The issue already has an assignee.

**Comment by korthout** [comment]
This issue was raised again recently in `#inc-cluster-with-zeebeprocessinglag`, so I've spend 10 minutes to build a summary of where things stand. I also came up with an idea to limit the impact of many due timers.

### Root cause

When many timers become due simultaneously (or in a short time frame), the `DueDateCheckScheduler` overwhelms the processing thread. This blocks other work (heartbeats, command processing) and can make partitions unhealthy. The problem also generalizes to job timeouts and message TTL.

### What has been fixed

Three sub-problems identified in the original investigation have all been resolved:

| Problem | Fix |
|---|---|
| Checker rescheduled with negative delay (runs immediately) | #9236 ✅ |
| Checker iterates all due timers without yielding | #9238 ✅ |
| Duplicate `TRIGGER` commands written for the same timer | #17128 ✅ (via scheduled command cache) |

### What remains

Despite these fixes, the core problem persists: **a large volume of *distinct* due timers can still overwhelm the logstream and block processing.** The fixes prevent duplicates and ensure yielding, but if there are genuinely many timers due (e.g. thousands/millions scheduled for the same time), the system still produces a massive backlog of trigger commands that the processor can't keep up with.

### Possible next step

One viable approach: **use the existing scheduled command cache to throttle the number of in-flight trigger commands.** The cache already tracks which `TRIGGER` commands have been written but not yet processed. We could stop writing new trigger commands once the in-flight count exceeds a threshold. This would create natural backpressure — as triggers are processed and removed from the cache, more can be written.

Note that this differs from the checker yielding after a certain number of commands send, as in that case it would continue sending more on the next run. The approach I'm suggesting now is to limit the total number of inflight TRIGGER commands.

The limit would need to be **dynamically adjusted**. Too low and timer processing unnecessarily slows down under normal load, too high and it fails to prevent the pathological case. Two candidate signals for adjusting the limit:

- **Command cache drain rate:** directly observe how quickly entries are removed from the cache (i.e., how fast triggers are being processed). If the cache drains quickly, allow more in-flight triggers. If it stalls, reduce the limit.
- **Existing backpressure state:** if the flow control system is already under pressure (low limit, high rejection rate), that's an indirect indicator of system overload. The timer throttle could read this state and reduce its own limit accordingly.

Further caveats: the `BoundedCommandCache` has a fixed capacity (100k) with random eviction, so the threshold must stay well below that. The cache size isn't currently exposed through the `ScheduledCommandCache` interface, so some plumbing is needed. The same approach could generalize to job timeouts and message TTL.

### Alternative next steps

Some other options to explore (but that do not resolve the root cause) are:
- allow to temporarily disable all timer checkers, or for a specific element id or so, e.g. through the management API
- consider banned instances in timer checker

**Comment by RobertDorlan** [comment]
Successfully tested agains 8.-SNAPSHOT


As a user I'm not able to completely f** up my cluster, or at least there is some way to recover from this. ✅ 


<img width="1728" height="929" alt="Image" src="https://github.com/user-attachments/assets/5910addf-d3e0-4a13-9e9e-a5ea6f1fc183" />




NOTE: We also have now preventions of "simple" loops

<img width="1262" height="742" alt="Image" src="https://github.com/user-attachments/assets/58b7cd30-6833-45c9-b06a-9e17d769e089" />

**Comment by korthout** [comment]
@RobertDorlan Can you share a bit more about the test you ran?
- It appears there was practically no processing happening on the cluster, were there really millions of timers triggering?
- It appears the process deployment was rejected, did you deploy and run another?

EDIT: I'm surprised this was not reproducible as the issue was raised again recently in `#inc-cluster-with-zeebeprocessinglag`

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** None  
**To:** Mar M2 (8.9)  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by RobertDorlan** [comment]
Hey @korthout you were right I have to say. 
Please have a look and let me know if I should do any other test.
BTW Thanks for your help. (I really appreciate it) 
I re-do my testing agains **8.9-SNAPSHOT**

These are my new test and findings: 


**Test 1:** 
Deployed the following bpmn:
[testInc8991.bpmn.txt](https://github.com/user-attachments/files/26121592/testInc8991.bpmn.txt)

this BPMN do the following
Inside each iteration:

A user task (“Stop waiting for the timer”) is created.
At the same time, a non-interrupting timer triggers every 5 seconds repeatedly.
Either:
The user completes the task → process ends for that instance, or
The timer fires → it also leads to the end (without stopping the task, since it’s non-interrupting).
So overall:
It creates 1000 user tasks, each with a repeating 5-second timer that can trigger alongside the task until completion.

FINDINGS
- The first 1000 records where processed quite fast
- We can see a peak in the processing per partition and the processing queue size.(but after 10 minutest it stabilise)

<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/a71d2759-6fb2-40bc-9d56-dbe7560fcd5b" />

<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/82ca7990-c5e0-49e6-b61c-9c52c761bd4c" />

<img width="862" height="304" alt="Image" src="https://github.com/user-attachments/assets/43626ae0-2642-4adb-ade8-436fea891737" />




**Test 2:** 
Deployed and run the same bpmn

This time It creates 3000 user tasks, each with a repeating 5-second timer that can trigger alongside the task until completion.

FINDINGS
- This time the records where processed slower (and everything stop when the Partition 1 goes Unhealthy) 
 - We can see a trough in processing per partition as Partition 1 goes Unhealthy (After 1 hour the Partition is Healthy) But not deployment can be done 
- We can see a peak on processing queue size 
- Backpressure is at 99.5%


<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/6ed38157-73cf-4539-bc73-b8f55217ab32" />

<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/5a698b36-2d66-46fb-9a05-2ab860eea654" />

<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/5903d29a-780d-48ce-a1c1-05f933117d0b" />

<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/525e6936-24ac-40c0-b70c-901e2200933b" />

<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/79fe4fa2-dbe3-4ce0-a486-aea0373a6c64" />

<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/55046e93-90c7-4020-9bd5-31b231b32d11" />

Also getting the following error when trying to access Operate (After 3-6 minute we can access Operate) 
<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/e4bc3e06-9a4f-4ed8-8e5b-2ce69e252bd8" />





**Test 3:** 
Deployed and run 30 times the same bpmn, the first 20 iterating 10 times and the last 10 iterating 3000 times

FINDINGS
- Partition 1, 2 and 3 went Unhealthy after 5 minutes
<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/7604b99c-cf52-43a4-811f-774ec1a90b52" />

- Backpresure at 88% after 5 minutes 
- Backpresure at 95% after 10 minutes 
- Backpresure at 99.7% after 15 minutes 

<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/6c8341ec-2ab7-454f-abaf-5824ed6702cd" />

- Not able to do any Deployment after 15 minutes
<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/2f1d63f4-be3e-41b4-b713-3114d3be2d11" />

<img width="862" height="521" alt="Image" src="https://github.com/user-attachments/assets/cb370306-7474-44cb-b621-3e7a7ebc76a5" />

- Not able to do any Deployment after 1 hour



**Test4:**

Tried to start 500 process instances with the same bpmn and got the following message after 199 started successfully: 

{"type":"about:blank","title":"DEADLINE_EXCEEDED","status":504,"detail":"Expected to handle request, but request timed out between gateway and broker","instance":"/3927af64-d168-438d-ba8e-1b01de6f34aa/v2/process-instances"}

7 minutes lated all partitions were Unhealthy

**Comment by korthout** [comment]
Alright. Thanks @RobertDorlan for testing.

Let me get started on this bug.

**Comment by korthout** [comment]
I'm planning to work on this issue in the upcoming iteration.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** May - M1  
**To:** May - M2  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** May - M2  
**To:** June - M1  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by korthout** [comment]
@ChrisKujawa I'm actually working on this, but it got closed by automation on Reliability testing side.

**Comment by ChrisKujawa** [comment]
@korthout sorry idk something went off yesterday with the gh actions and moved several issues on our board. I tried to clean up this in the morning an accidently moved it to done. Looks like this closed it.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** June - M1  
**To:** June - M2  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** June - M2  
**To:** July - M1  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** July - M1  
**To:** July - M2  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by korthout** [comment]
While I have work in progress, I need to prioritize other items at the moment. I plan to come back to it in the next iteration.

**Comment by korthout** [comment]
I'm dropping the likelihood. It's not something we've seen happen in production, and has only been reproduced under specific situations. The impact remains severe, so we need to resolve it, but it's not a release blocker.
